File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.cpp
#include <string>
#include <vector>
#include <deque>
#include <algorithm>

using namespace std;

int solution(int cacheSize, vector<string> cities) {
    int answer = 0;
    
    deque<string> cache;
    
    for(string city : cities) {
        
        transform(city.begin(), city.end(), city.begin(), ::tolower);
        
        bool hit = false;
        int idx = 0;
        while(idx < cache.size()) {
            if(cache[idx] == city) {
                hit = true;
                break;
            }
            idx++;
        }
        
        cache.push_back(city);
        
        if(hit) {
            answer++;
            cache.erase(cache.begin() + idx);
        }
        else {
            answer += 5;
            
            if(cache.size() > cacheSize) {
                cache.pop_front();
            }
        }        
    }
    
    return answer;
}

File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.py
from collections import deque

def solution(cacheSize, cities):
    answer = 0
    
    if cacheSize == 0:
        return 5 * len(cities)
    
    cache = deque()
    
    for city in cities:
        city = city.lower()
        
        if city in cache:
            answer += 1
            cache.remove(city)
            cache.append(city)
        else:
            answer += 5
            if len(cache) < cacheSize:
                cache.append(city)
            else:
                cache.popleft()
                cache.append(city)
    
    return answer